### PR TITLE
[mini] Add optional method name to --stats

### PR DESCRIPTION
**Pull request by rgarc**
Implement parsing for optional method name to `--stats` command line argument.

### CODE REVIEW

**Comment by lambdageek** [comment]
@monojenkins build

File: mono/mini/driver.c [hunk 1]
```diff
@@ -1745,6 +1757,8 @@ mono_jit_parse_options (int argc, char * argv[])
 			mono_counters_enable (-1);
 			mono_atomic_store_bool (&mono_stats.enabled, TRUE);
 			mono_atomic_store_bool (&mono_jit_stats.enabled, TRUE);
+		} else if (strncmp (argv [i], "--stats=", 8) == 0) {
+			handle_stats_name (argv [i]);
```
Comment (CoffeeFlux) [inline mono/mini/driver.c]: If it hits this condition, it means it has *not* run the block for `strcmp (argv [i], "--stats") == 0`. While you will set the method name, you won't enable the counters or stat collection.

File: mono/mini/driver.c [hunk 2]
```diff
@@ -1689,6 +1689,18 @@ mono_get_version_info (void)
 
 static gboolean enable_debugging;
 
+void
+handle_stats_name (char *arg)
```
Comment (CoffeeFlux) [inline mono/mini/driver.c]: This should be static, as it's not used outside of this file. I'd also make the argument name something more useful than `arg` if you intend to be passing only the method name.

File: mono/mini/driver.c [hunk 3]
```diff
@@ -1689,6 +1689,18 @@ mono_get_version_info (void)
...
+{
+	size_t method_name_len = strlen (arg);
+	if (method_name_len == 0) {
```
Comment (CoffeeFlux) [inline mono/mini/driver.c]: This will never be zero because you're comparing the full argv, including the `--stats=` portion. I recommend passing only the method name itself to this function.

File: mono/mini/mini-runtime.c [hunk 4]
```diff
@@ -104,6 +104,7 @@
 
 static guint32 default_opt = 0;
 static gboolean default_opt_set = FALSE;
+static char *stats_method_name = NULL;
```
Comment (CoffeeFlux) [inline mono/mini/mini-runtime.c]: I think you'd be better off not making this static, and just using an extern in the header. You also won't need the helper methods in that case. See `sdb_options` in mini-runtime for an example.

File: mono/mini/mini-runtime.c [hunk 5]
```diff
@@ -4749,6 +4771,7 @@ mini_cleanup (MonoDomain *domain)
 {
 	print_jit_stats ();
 	mono_counters_dump (MONO_COUNTER_SECTION_MASK | MONO_COUNTER_MONOTONIC, stdout);
+	mono_stats_method_name_clean();
```
Comment (CoffeeFlux) [inline mono/mini/mini-runtime.c]: Don't worry about freeing it here. If you were to free it, keep in mind that calling `g_free` on a pointer will not change the value of the variable that holds that pointer, so you would have a dangling reference to freed memory. A later attempt to access that would cause a crash.

File: mono/mini/driver.c [hunk 6]
```diff
@@ -1689,6 +1689,18 @@ mono_get_version_info (void)
...
+		fprintf (stderr, "Missing method name in --stats command line option\n");
+		exit (1);
+	}
+	// 8 = strlen("--stats=")
+	mono_stats_set_method_name (g_strndup (&arg [8], method_name_len));
```
Comment (CoffeeFlux) [inline mono/mini/driver.c]: As-is, this will duplicate too much since you're getting the length of the full string but then only duplicating starting at index 8. Passing only the method name makes this problem go away as well.

File: mono/mini/driver.c [hunk 7]
```diff
@@ -2190,6 +2204,8 @@ mono_main (int argc, char* argv[])
 			mono_counters_enable (-1);
 			mono_atomic_store_bool (&mono_stats.enabled, TRUE);
 			mono_atomic_store_bool (&mono_jit_stats.enabled, TRUE);
+		} else if (strncmp (argv [i], "--stats=", 8) == 0) {
```
Comment (CoffeeFlux) [inline mono/mini/driver.c]: See above.

File: mono/mini/driver.c (hunk 2 above)
Comment (CoffeeFlux) [reply mono/mini/driver.c]: Oh, another nit: I'd also call it something like `set_stats_method_name` - handle is unnecessarily vague.

**Comment by vargaz** [comment]
What is the aim of this PR ?

**Comment by lambdageek** [comment]
@vargaz This is the first commit in the intern project.  (Note the code is going to a separate branch, not to master).
The eventual goal is to add a `mono --stats=MethodSpec` option that can be used with the JIT and the interpreter to dump stats just before the specified method is executed.  The idea is that for XA we could use this to measure how much work the runtime and frameworks are doing before user code starts running.

File: mono/mini/driver.c [hunk 8]
```diff
@@ -2186,10 +2199,12 @@ mono_main (int argc, char* argv[])
 			mono_jit_set_aot_mode (MONO_AOT_MODE_LLVMONLY_INTERP);
 		} else if (strcmp (argv [i], "--print-vtable") == 0) {
 			mono_print_vtable = TRUE;
-		} else if (strcmp (argv [i], "--stats") == 0) {
+		} else if (strncmp (argv [i], "--stats", 7) == 0) {
 			mono_counters_enable (-1);
 			mono_atomic_store_bool (&mono_stats.enabled, TRUE);
 			mono_atomic_store_bool (&mono_jit_stats.enabled, TRUE);
+			if (argv [i][7] == '=') 
+				set_stats_method_name (&argv [i][8]);
```
Comment (CoffeeFlux) [inline mono/mini/driver.c]: See above.

File: mono/mini/driver.c [hunk 9]
```diff
@@ -1741,10 +1752,12 @@ mono_jit_parse_options (int argc, char * argv[])
 			MonoDebugOptions *opt = mini_get_debug_options ();
 
 			opt->break_on_exc = TRUE;
-		} else if (strcmp (argv [i], "--stats") == 0) {
+		} else if (strncmp (argv [i], "--stats", 7) == 0) {
 			mono_counters_enable (-1);
 			mono_atomic_store_bool (&mono_stats.enabled, TRUE);
 			mono_atomic_store_bool (&mono_jit_stats.enabled, TRUE);
+			if (argv [i][7] == '=')
+				set_stats_method_name (&argv [i][8]);
```
Comment (CoffeeFlux) [inline mono/mini/driver.c]: You can just do `argv [i] + 8` and avoid the dereferencing dance.

File: mono/mini/mini-runtime.h [hunk 10]
```diff
@@ -422,10 +422,11 @@ MONO_API void        mono_parse_env_options         (int *ref_argc, char **ref_a
 MONO_API char       *mono_parse_options_from        (const char *options, int *ref_argc, char **ref_argv []);
 MONO_API int         mono_regression_test_step      (int verbose_level, const char *image, const char *method_name);
 
-
```
Comment (CoffeeFlux) [inline mono/mini/mini-runtime.h]: Unrelated whitespace change. Generally avoid these in diffs.

File: mono/mini/mini-runtime.h [hunk 11]
```diff
@@ -422,10 +422,11 @@ MONO_API void        mono_parse_env_options         (int *ref_argc, char **ref_a
...
 void                   mono_interp_stub_init         (void);
 void                   mini_install_interp_callbacks (const MonoEECallbacks *cbs);
 
+extern char           *stats_method_name;
```
Comment (CoffeeFlux) [inline mono/mini/mini-runtime.h]: I would put this right under `sdb_options` with most of the externs. `mono_interp_callbacks_pointer` is only separated to group it with `mini_install_interp_callbacks`, but there's no reason for `stats_method_name` to live here.

File: mono/mini/driver.c [hunk 12]
```diff
@@ -2186,10 +2202,12 @@ mono_main (int argc, char* argv[])
 			mono_jit_set_aot_mode (MONO_AOT_MODE_LLVMONLY_INTERP);
 		} else if (strcmp (argv [i], "--print-vtable") == 0) {
 			mono_print_vtable = TRUE;
-		} else if (strcmp (argv [i], "--stats") == 0) {
+		} else if (strncmp (argv [i], "--stats", 7) == 0) {
```
Comment (CoffeeFlux) [inline mono/mini/driver.c]: I would actually prefer you keep this as a separate condition. The problem with this code is that I could pass the argument `--statsASDF` and it would still trigger this block because you're using `strncmp`. The actual case where this matters is if we later were to add some other argument that begins with `--stats`: for example, `--stats-extra`.

File: mono/mini/driver.c [hunk 13]
```diff
@@ -1689,6 +1689,28 @@ mono_get_version_info (void)
...
+		exit (1);
+	}
+	if (stats_method_name != NULL) {
+		g_free (stats_method_name);
+	}
+	stats_method_name = g_strndup (name, method_name_len);
+}
+
+static void
+mono_setup_counters (void)
```
Comment (CoffeeFlux) [inline mono/mini/driver.c]: Nit: I'd go with `mono_enable_counters`.

File: mono/mini/mini-runtime.h (hunk 10 above)
Comment (CoffeeFlux) [reply mono/mini/mini-runtime.h]: This is still not resolved. You should check the actual diff before marking anything.

File: mono/mini/mini-runtime.h [hunk 14]
```diff
@@ -422,10 +423,10 @@ MONO_API void        mono_parse_env_options         (int *ref_argc, char **ref_a
 MONO_API char       *mono_parse_options_from        (const char *options, int *ref_argc, char **ref_argv []);
 MONO_API int         mono_regression_test_step      (int verbose_level, const char *image, const char *method_name);
 
-
 void                   mono_interp_stub_init         (void);
 void                   mini_install_interp_callbacks (const MonoEECallbacks *cbs);
 
+
```
Comment (CoffeeFlux) [inline mono/mini/mini-runtime.h]: Unrelated whitespace change.

File: mono/mini/mini-runtime.c [hunk 15]
```diff
@@ -104,6 +104,7 @@
 
 static guint32 default_opt = 0;
 static gboolean default_opt_set = FALSE;
+char *stats_method_name = NULL;
```
Comment (CoffeeFlux) [inline mono/mini/mini-runtime.c]: I'd at least have a newline between this and the statics; they don't seem to be related.

File: mono/mini/mini-runtime.h (hunk 14 above)
Comment (rgarc) [reply mono/mini/mini-runtime.h]: I got confused with the white space because I was always comparing to the most recent commit. I fixed it by `git diff` 'ing with `0a51d89c4238c46f07f8b5405383547c1530809b` which was the commit before my first one. I restored the whitespace back to what it was in the aforementioned commit.

Comment (CoffeeFlux) [reply mono/mini/mini-runtime.h]: The Github preview of the diff shows the overall change to the file, see the Files Changed tab. It looks to be like a newline is being added that wasn't there before this PR.

File: mono/mini/mini-runtime.h [hunk 16]
```diff
@@ -423,10 +423,10 @@ MONO_API void        mono_parse_env_options         (int *ref_argc, char **ref_a
 MONO_API char       *mono_parse_options_from        (const char *options, int *ref_argc, char **ref_argv []);
 MONO_API int         mono_regression_test_step      (int verbose_level, const char *image, const char *method_name);
 
+
 void                   mono_interp_stub_init         (void);
 void                   mini_install_interp_callbacks (const MonoEECallbacks *cbs);
 
-
```
Comment (rgarc) [inline mono/mini/mini-runtime.h]: We attempted to restore the whitespace to how we found it originally.

**Comment by CoffeeFlux** [comment]
@monojenkins build

**Comment by CoffeeFlux** [comment]
@monojenkins build failed
